src/scanner/mod.rs

```diff
@@ -87,54 +98,72 @@ impl Scanner {
         open_sockets
     }
 
-    /// Given a port, scan it.
+    /// Given a port, scan it self.tries times.
     /// Turns the address into a SocketAddr
     /// Deals with the <result> type
     /// If it experiences error ErrorKind::Other then too many files are open and it Panics!
-    /// ese any other error, it returns the error in Result as a string
-    /// If no  errors occur, it returns the port number in Result to signify the port is open.
+    /// If any other error, it returns the error in Result as a string
+    /// If no errors occur, it returns the port number in Result to signify the port is open.
     /// This function mainly deals with the logic of Results handling.
     /// # Example
     ///
     ///     self.scan_port(10:u16)
     ///
     /// Note: `self` must contain `self.ip`.
     async fn scan_socket(&self, socket: SocketAddr) -> io::Result<SocketAddr> {
-        match self.connect(socket).await {
-            Ok(x) => {
-                debug!(
-                    "Connection was successful, shutting down stream {}",
-                    &socket
-                );
-                match x.shutdown(Shutdown::Both) {
-                    Err(e) => debug!("Shutdown stream error {}", &e),
-                    _ => {}
-                }
-                if !self.greppable {
-                    if self.accessible {
-                        println!("Open {}", socket.to_string());
-                    } else {
-                        println!("Open {}", socket.to_string().purple());
+        debug!("self.tries: {}", self.tries);
+        for nr_try in 1..=self.tries {
+            debug!("Try number: {}", nr_try);
+
+            match self.connect(socket).await {
+                Ok(x) => {
+                    debug!(
+                        "Connection was successful, shutting down stream {}",
+                        &socket
+                    );
+                    match x.shutdown(Shutdown::Both) {
+                        Err(e) => debug!("Shutdown stream error {}", &e),
+                        _ => {}
                     }
+                    if !self.greppable {
+                        if self.accessible {
+                            println!("Open {}", socket.to_string());
+                        } else {
+                            println!("Open {}", socket.to_string().purple());
+                        }
+                    }
+
+                    debug!("Return Ok after {} tries", nr_try);
+                    return Ok(socket);
                 }
+                Err(e) => {
+                    let error = match e.kind() {
```

Comment:
This is already way nicer, but I think it can be even more tight. This is what I propose:

```rust
let error_string = e.to_string();
if e.kind() == ErrorKind::Other {
    debug!("Socket connect error: {} {}", &error_string, &socket);
    if !error_string.contains("No route to host")
        && !error_string.contains("Network is unreachable")
    {
        panic!("Too many open files. Please reduce batch size. The default is 5000. Try -b 2500.");
    }
}

if nr_try == self.tries {
    debug!("Return Err after {} tries", self.tries);
    return Err(io::Error::new(io::ErrorKind::Other, error_string));
}
```

This avoids the creation a new `io::Error` if it is not needed

Reply:
I add this on my local version.